File: coef.c
#include <stdio.h>
#include <math.h>

// É PRA SER um código pra achar o coeficiente de correlação entre n variáveis inseridas pelo user. 
// Como não tenho tanto conhecimento assim de estatística, não testei o suficiente pra saber se está certo ainda

double mean(double values[], int size) {
double sum = 0;
for (int i = 0; i < size; i++) {
sum += values[i];
}
return sum / size;
}

double variance(double values[], int size) {
double meanValue = mean(values, size);
double sum = 0;
for (int i = 0; i < size; i++) {
sum += pow(values[i] - meanValue, 2);
}
return sum / (size - 1);
}

double covariance(double x[], double y[], int size) {
double meanX = mean(x, size);
double meanY = mean(y, size);
double sum = 0;
for (int i = 0; i < size; i++) {
sum += (x[i] - meanX) * (y[i] - meanY);
}
return sum / (size - 1);
}

double correlationCoefficient(double x[], double y[], int size) {
double covarianceXY = covariance(x, y, size);
double varianceX = variance(x, size);
double varianceY = variance(y, size);
return covarianceXY / sqrt(varianceX * varianceY);
}

int main() {
int size;
printf("Insira o número de elementos: ");
scanf("%d", &size);

double x[size], y[size];
for (int i = 0; i < size; i++) {
    printf("Insira o valor de x[%d]: ", i);
    scanf("%lf", &x[i]);
    printf("Insira o valor de y[%d]: ", i);
    scanf("%lf", &y[i]);
}

double coefficient = correlationCoefficient(x, y, size);
printf("O coeficiente de correlação linear é: %lf\n", coefficient);

return 0;
}

File: coef.py
import numpy as np

#versao em python

def coef_correlation(x, y):
    x_mean = np.mean(x)
    y_mean = np.mean(y)
    x_dev = x - x_mean
    y_dev = y - y_mean
    return np.sum(x_dev * y_dev) / np.sqrt(np.sum(x_dev**2) * np.sum(y_dev**2))

if __name__ == "__main__":
    n = int(input("Insira o número de elementos: "))
    x = np.zeros(n)
    y = np.zeros(n)
    for i in range(n):
        x[i] = float(input(f"Insira o valor x[{i}]: "))
        y[i] = float(input(f"Insira o valor y[{i}]: "))
    print("O coeficiente de correlação linear é:", coef_correlation(x, y))
